### PR TITLE
feat: new high availability doc for Dqlite

### PR DESCRIPTION
**Pull request by manadart**
This venerable document, out of date even for Mongo, now reflects HA in the Dqlite world.

## QA steps

None required.

## Documentation changes

This is one.

## Links

**Jira card:** [JUJU-4997](https://warthogs.atlassian.net/browse/JUJU-4997)

### CODE REVIEW

File: doc/high_availability.md [hunk 1]
```diff
@@ -1,110 +1,74 @@
...
-state and choose whether to run the singleton workers again.
-
-It also means that we can never accidentally have two
-singleton workers performing operations at the same time.
+HA mode is invoked via the `enable-ha` command. By default, this will ensure 
+three controllers, but using the `-n` flag will allow running with five or 
+seven controllers.
+
+The number of controllers can be reduced by invoking `remove-machine` with a 
+controller machine ID, and increased by re-running the `enable-ha` command.
```
Comment (jameinel) [inline doc/high_availability.md:12]: I certainly think that long term this will be nice to become `juju add-unit -m controller juju-controller` and/or `juju remove-unit`.
Someday :)

File: doc/high_availability.md [hunk 2]
```diff
@@ -1,110 +1,74 @@
...
+When starting Dqlite, the worker must bind it to an IP address on the local
+machine. It is a requirement that there is a unique local-cloud-scoped address
+for Dqlite to use. If there is no unique address, new nodes will not be joined 
+to the cluster until one can be determined. See _Controller Charm_ below.
+
+Beyond joining nodes to the cluster, Juju does not manage Dqlite node roles.
+This is handled within Dqlite itself. A cluster will have one leader, voters
+which are eligible to participate in leader elections, stand-bys and spares.
+
+Juju does not predicate any logic on node roles.
```
Comment (jameinel) [inline doc/high_availability.md:30]: It does do logic based on who the leader is, but I don't think we do any other logic.
I'm not sure if that subtlety is worthwhile.

File: doc/high_availability.md [hunk 3]
```diff
@@ -1,110 +1,74 @@
...
+
+If the number of controller instances is reduced to one, the `dbaccessor` 
+worker will recognise this scenario and proactively reconfigure the cluster to
+be constituted by the local node only.
```
Comment (jameinel) [inline doc/high_availability.md:34]: How do we handle HA=2? Do we keep the second one as a hot spare that can't vote itself to primary, or do we drop it lower than that?
My original Mongo logic made it so that you would have Availability over Correctness (as long as the secondary went down, you would still have availability, but I realize it actually means you lose redundancy, and it *might* be better to lose availability so that nothing can be done that won't at least be replicated somewhere.)

File: doc/high_availability.md [hunk 4]
```diff
@@ -1,110 +1,74 @@
...
+
+When machines in the control plane change, agent configuration files are 
+re-written with usable API addresses from all controllers. Agents will try
+these addresses in random order when connecting, so that an inaccessible 
+controller just results in connection to another.
+
+The list of addresses supplied to agent configuration can be influenced by the
+`juju-ctrl-space` controller configuration article. This is supplied with a
+space name in order that agent-controller communication can be isolated to 
+specific networks.
```
Comment (jameinel) [inline doc/high_availability.md:54]: I'm pretty sure this is still spelled `juju-mgmt-space` in 4.0. Ideally it would also become a binding rather than a controller configuration. But since we don't model the relations of each agent to the controller, it doesn't really have a place to live. It *is* what `extra-bindings` was designed for, but I know we don't like the nomenclature. With the caveat that all progress on *improving* the network model has stalled, which means we at least have *something* that can be used, even if it isn't perfect yet.

https://github.com/juju/juju/blob/d0d39e6e8ee41039b1befa728c03db19b6cbd4ab/controller/config.go#L240

File: doc/high_availability.md [hunk 5]
```diff
@@ -1,110 +1,74 @@
...
+
+### API Addresses for Clients
+
+Each time the Juju client establishes a connection to the Juju controller, it
+is sent the current list of API addresses and updates these in the local store. 
+The client's first connection attempt is always to the last address that it 
+used successfully. Others are tried subsequently if required.
+
+Addresses used by clients are not influenced by the `juju-ctrl-space`
```
Comment (jameinel) [inline doc/high_availability.md:63]:
```suggestion
Addresses used by clients are not influenced by the `juju-mgmt-space` 
```

File: doc/high_availability.md (hunk 1 above)
Comment (wallyworld) [reply doc/high_availability.md:12]: For 4.0, I really was hoping we'd do this TBH, especially now the controller charm does what it's supposed to do wrt configuring dqlite etc. Is there a concrete reason why we can't now drop enable-ha in favour of add/remove unit?

File: doc/high_availability.md [hunk 6]
```diff
@@ -1,110 +1,74 @@
...
+to the cluster until one can be determined. See _Controller Charm_ below.
+
+Beyond joining nodes to the cluster, Juju does not manage Dqlite node roles.
+This is handled within Dqlite itself. A cluster will have one leader, voters
+which are eligible to participate in leader elections, stand-bys and spares.
+
+Juju does not predicate any logic on node roles.
+
+If the number of controller instances is reduced to one, the `dbaccessor` 
+worker will recognise this scenario and proactively reconfigure the cluster to
```
Comment (gfouillet) [inline doc/high_availability.md:33]: **suggestion:** I think `recognise` should be in the present tense.
```suggestion
worker recognises this scenario and proactively reconfigure the cluster to
```

References: https://www.enago.com/academy/ifthen-using-conditional-sentences-in-academic-writing/
(we are in a case where the **Zero conditional sentences** case => there is no assumption on the behavior of `dbaccessor` in the case of the number of controller instances drop to one. It is a known, sure consequences (at least if we have no bug :) )

**suggestion:** maybe search `will` usages and assess it regarding above rules ;)

File: doc/high_availability.md [hunk 7]
```diff
@@ -1,110 +1,74 @@
...
+configuration.
+
+### Singular Workers
+
+Many workers, such as the `dbaccessor` worker above, run on all controller 
+nodes, but there are some workers that must have exactly one instance running. 
+An obvious example of this is a model's compute provisioner - we don't want 
+multiple actors attempting to start a cloud instance for a new machine.
+
+The controller that such singular workers run on is determined by the lease
```
Comment (gfouillet) [inline doc/high_availability.md:73]: **suggestion:**  If this documentation exists, we should have a link to it. If not, we should add a comment

```md
<!-- TODO: add a link to lease subsystem documentation whenever it lands -->
```

**Review by gfouillet — COMMENTED** [review]
Very interesting. However, I have some thought and suggestion to improve clarity.

File: doc/high_availability.md [hunk 8]
```diff
@@ -1,110 +1,74 @@
 High Availability (HA)
 ======================
```
Comment (tmihoc) [inline doc/high_availability.md:2]: Based on the content that follows, the topic here is just _controller_ high availability, so I'd update the title to that.

File: doc/high_availability.md (hunk 7 above)
Comment (tmihoc) [reply doc/high_availability.md:73]: Some notes and a possible solution:

(1) such as the `dbaccessor` worker above

This sort of reference usually signals an information architecture problem. 

(2) "run on all controller nodes"

Below we also speak of "instance". And I'm guessing we really mean "unit". Whichever we pick, I'd be consistent and maybe go for "unit" -- in Juju, horizontal scaling and, piggybacking on that, HA, is a matter of units. (Even if all labels apply, the inconsistency is causing me to have to think harder than strictly necessary.)

(3) "singular workers" 

Seems like unnecessary jargon.

(4) "must have exactly one instance running" 

Confusing -- what could it even mean in the context of HA? I'd rephrase as "run on exactly one controller unit, where the unit is determined by the lease sub-system."

Overall suggestion:
Have a section "Workers involved in controller HA" or something like that, then mention both the `dbaccessor` worker and others, then comment (this can even be a subsection -- even if the content is just one line, it's logically a subtopic) on how they differ with respect to whether they run on all controller units or on exactly one controller unit, as determined by the lease sub-system.

File: doc/high_availability.md (hunk 1 above)
Comment (manadart) [reply doc/high_availability.md:12]: Other than hoping to get it done one day, no.

File: doc/high_availability.md (hunk 3 above)
Comment (manadart) [reply doc/high_availability.md]: Today, if 1 machine of 3 goes away, it is still recorded in _cluster.yaml_ as a voter, so you rely on both the 2 remaining for quorum.

If you go to 1, we invoke backstop and reconfigure. If you go back to HA-3, the removed node is still in _cluster.yaml_, but becomes a _spare_.

File: doc/high_availability.md [hunk 9]
```diff
@@ -1,110 +1,74 @@
...
-master changes, the machine agent will reconnect to the
-state and choose whether to run the singleton workers again.
-
-It also means that we can never accidentally have two
-singleton workers performing operations at the same time.
+HA mode is invoked via the `enable-ha` command. By default, this will ensure 
+three controllers, but using the `-n` flag will allow running with five or 
+seven controllers.
+
+The number of controllers can be reduced by invoking `remove-machine` with a
```
[the remark this answers is not in the document]
Comment (manadart) [reply doc/high_availability.md:11]: Regarding the first 3 points, your understanding is correct.

Regarding the latter 3:
- Juju will run with whatever number of controllers, you will just have degraded resilience.
- `enable-ha` does validate that you have an odd number of controllers >1 && <=7. This document is for the current behaviour, not ideal behaviour.

File: doc/high_availability.md [hunk 10]
```diff
@@ -1,110 +1,74 @@
...
+The number of controllers can be reduced by invoking `remove-machine` with a 
+controller machine ID, and increased by re-running the `enable-ha` command.
+
+### Dqlite
+
+Each controller is a Dqlite node. The `dbaccessor` worker on each controller is 
+responsible for maintaining the Dqlite cluster. When entering HA mode, the 
+`dbaccessor` worker will configure the local Dqlite node as a member of the 
+cluster.
+
```
[the remark this answers is not in the document]
Comment (manadart) [reply doc/high_availability.md:20]: I think I've addressed this sufficiently in the latest iteration.

File: doc/high_availability.md (hunk 2 above)
Comment (manadart) [reply doc/high_availability.md:30]: Dropped it.

**Review by tmihoc — APPROVED** [review]
Thanks for all the changes! Left a few more comments. Will pre-approve, because I think this is a valuable addition as is, but please see those as well, as the result can be clearer.

File: doc/high_availability.md [hunk 11]
```diff
@@ -1,110 +1,76 @@
...
+machine.
+
+Single instance workers are those declared in the model manifolds configuration
+that use the `isResponsible` decorator. This in turn is based on a flag set by the
+`singular` worker.
+
+The `singular` worker only sets the flag if it is the current lease holder for 
+the `singular-controller` namespace. See the appropriate documentation for more 
+information on leases.
+
```
[the remark this answers is not in the document]
Comment (manadart) [reply doc/high_availability.md:74]: The lease _manager_ has some nominal documentation, but there is no documentation dedicated to the lease system as a concept. This is a gap.
